[NCCL] Changed FutureNCCL's then callback logic for better efficiency.

sinannasir opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43310 DDP communication hook examples
* **#42869 [NCCL] Changed FutureNCCL's then callback logic for better efficiency.**

We realized that when we invoke a simple callback that divides the tensors by `world_size` after `allreduce`, the performance was almost 50% lower in terms of QPS compared to the case where a simple `allreduce` hook is used with no `then` callback.

The main problem was as we call `work.wait()` before invoking `then` callback, we were synchronizing `work`'s stream with the default PyTorch stream inside [`runHook`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L609) and stalling the backward computation.

In that PR, we ensure that FutureNCCL's `then` callback is not stalling the backward computation. Assuming single-process single-device, `FutureNCCL` gets a new stream from device's pool using `at::cuda::getStreamFromPool` to run `callback` and before invoking the `callback` inline it synchronizes `WorkNCCL`'s stream by callback's stream not the default stream.

Differential Revision: [D23055807](https://our.internmc.facebook.com/intern/diff/D23055807/)

sinannasir committed:
[NCCL] Changed FutureNCCL's then callback logic for better efficiency.

We realized that when we invoke a simple callback that divides the tensors by `world_size` after `allreduce`, the performance was almost 50% lower in terms of QPS compared to the case where a simple `allreduce` hook is used with no `then` callback.

@pritam realized that we should run the callback in `then` on a different stream and synchronize the NCCL stream to that stream instead and not the default device stream. Now a new `FutureNCCL` is returned from `then` hold the return value of the callback and new cudaEvents that recorded the stream that guards this callback.. This new FutureNCCL actually synchronizes on the new stream we used for the callback when its wait() is called.

**Ongoing discussions:**
* Calling `fut.wait()` (instead of `fut.value()`) inside the callback of the allreduce then divide hook is okay.
* Calling `fut.wait()` if a new futureNCCL object is defined inside then callback is essential to get correct results. It could be fine in terms of performance since within the scope of then callback the default stream becomes the stream that guards the callback but not PyTorch's default device stream.

Differential Revision: [D23055807](https://our.internmc.facebook.com/intern/diff/D23055807/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] Changed FutureNCCL's then callback logic for better efficiency."


We realized that when we invoke a simple callback that divides the tensors by `world_size` after `allreduce`, the performance was almost 50% lower in terms of QPS compared to the case where a simple `allreduce` hook is used with no `then` callback.

@pritamdamania87 realized that we should run the callback in `then` on a different stream and synchronize the NCCL stream to that stream instead and not the default device stream. Now a new `FutureNCCL` is returned from `then` hold the return value of the callback and new cudaEvents that recorded the stream that guards this callback.. This new FutureNCCL actually synchronizes on the new stream we used for the callback when its wait() is called.

**Ongoing discussions:**
* Calling `fut.wait()` (instead of `fut.value()`) inside the callback of the allreduce then divide hook is okay.
* Calling `fut.wait()` if a new futureNCCL object is defined inside then callback is essential to get correct results. It could be fine in terms of performance since within the scope of then callback the default stream becomes the stream that guards the callback but not PyTorch's default device stream.

Differential Revision: [D23055807](https://our.internmc.facebook.com/intern/diff/D23055807/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -143,66 +144,140 @@ class ProcessGroupNCCL : public ProcessGroup {
+                try {
+                  fut = c10::make_intrusive<FutureNCCL>(
+                      cb(), deviceIndex_, thenFutCudaEvents);
+                } catch (std::exception& e) {
```
nit: `const std::exception`

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -143,66 +144,140 @@ class ProcessGroupNCCL : public ProcessGroup {
+          at::cuda::getStreamFromPool(deviceIndex_);
+      // Create a new cudaEvents object that will record callback's stream and
+      // used by the new FutureNCCL.
+      auto thenFutCudaEvents =
```
Any reason it needs to be a `std::vector` if it only holds one CUDA event?

sinannasir replied:
Referring the discussion in https://github.com/pytorch/pytorch/pull/42869#discussion_r470333493.

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -143,66 +144,140 @@ class ProcessGroupNCCL : public ProcessGroup {
+      if (ret != cudaSuccess && ret != cudaErrorNotReady) {
+        AT_CUDA_CHECK(ret);
+      }
+      if (ret == cudaErrorNotReady) {
```
nit: this block can be condensed to just `return ret != cudaErrorNotReady`

rohan-varma replied:
(or ret == cudaSuccess)

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -143,66 +144,140 @@ class ProcessGroupNCCL : public ProcessGroup {
   // wrapper to synchronize streams appropriately and it mostly enables
   // the async programming model of CUDA while trying to adhere to the
-  // Future interface.
+  // Future interface. FutureNCCL does not support NCCL_BLOCKING_WAIT flag
```
Can you clarify this a bit? If I do run with NCCL_BLOCKING_WAIT, does this mean that if I call .wait() on a future returned by get_future() it still won't block? Would we want to add this support later on (seems like we would), if so can we file a GH issue? Same goes for `barrier()` - what wouldn't work if we did barrier() with FutureNCCL?

sinannasir replied:
I think in case of `NCCL_BLOCKING_WAIT=0` it will just wait until streams are synchronized and it will return before the whole NCCL kernel completed on the GPU. 
My understanding is that in case of `barrier()` and `NCCL_BLOCKING_WAIT, then callback will become even more complicated. Any thoughts @pritamdamania87, do we wanna support those later?

rohan-varma reviewed (commented):
Nice! This overall looks great, have a few comments/questions inline.

rohan-varma commented:
Can we describe the perf issue with the previous solution and why this commit fixes it (basically no longer blocking default stream on allreduce in then(), which eliminated overlap)? Would be useful context for those who come across this in the future.

sinannasir committed:
Update on "[NCCL] Changed FutureNCCL's then callback logic for better efficiency."


We realized that when we invoke a simple callback that divides the tensors by `world_size` after `allreduce`, the performance was almost 50% lower in terms of QPS compared to the case where a simple `allreduce` hook is used with no `then` callback.

@pritamdamania87 realized that we should run the callback in `then` on a different stream and synchronize the NCCL stream to that stream instead and not the default device stream. Now a new `FutureNCCL` is returned from `then` hold the return value of the callback and new cudaEvents that recorded the stream that guards this callback.. This new FutureNCCL actually synchronizes on the new stream we used for the callback when its wait() is called.

**Ongoing discussions:**
* Calling `fut.wait()` (instead of `fut.value()`) inside the callback of the allreduce then divide hook is okay.
* Calling `fut.wait()` if a new futureNCCL object is defined inside then callback is essential to get correct results. It could be fine in terms of performance since within the scope of then callback the default stream becomes the stream that guards the callback but not PyTorch's default device stream.

Differential Revision: [D23055807](https://our.internmc.facebook.com/intern/diff/D23055807/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -772,10 +773,15 @@ std::shared_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
+      "WorkNCCL's getFuture API is only supported for single-process single-device mode.");
+  // Create a new FutureNCCL object after checking for single-process
+  // single-device mode.
+  std::shared_ptr<at::cuda::CUDAEvent> cudaEvent = {cudaEvents_,
+                                                    &(*cudaEvents_)[0]};
```
Why are we doing this and is this even correct? Shouldn't we just pass `cudaEvents_` to FutureNCCL?

sinannasir replied:
I was trying to address, @rohan-varma's comment https://github.com/pytorch/pytorch/pull/42869#discussion_r468863368. 

We just have one event, so probably using a vector isn't needed. I was just trying to create a pointer that shares the ownership with `cudaEvents_` and points to its first element. Since `at::cuda::CUDAEvent`'s constructor is deleted, I need to be careful here, but probably that's not an elegant solution and linter gives warning.

If using a vector is okay, I can just revert this and keep using vector inside `FutureNCCL`.

pritamdamania87 replied:
I think let's just pass a vector since I'm not sure if it's possible to create a shared_ptr pointing to one element of a shared_ptr vector.

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -734,21 +734,25 @@ They are used in specifying strategies for reduction collectives, e.g.,
+                will return another ``FutureNCCL`` that holds the return value of the callback and the
+                stream that runs the callback.
+
+                Note that ``fut.done()`` returns if work's NCCL streams were synchronized with PyTorch's
+                default device streams.
```
This doesn't seem to be right? In `completed()` we actually check if the work has completed on the GPU using `cudaEventQuery`.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -143,66 +144,129 @@ class ProcessGroupNCCL : public ProcessGroup {
-        std::shared_ptr<std::vector<at::Tensor>> outputs)
+        at::IValue value,
+        c10::DeviceIndex deviceIndex,
+        std::shared_ptr<at::cuda::CUDAEvent> cudaEvent)
```
In the constructor let's just assert the vector is of size 1.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -143,66 +144,129 @@ class ProcessGroupNCCL : public ProcessGroup {
+    // FutureNCCL's value is NCCL collective's outputs or callback() output with
+    // callbacks being invoked inline by addCallbackWithStream(), so
+    // markCompleted is not needed.
     void markCompleted(at::IValue /* unused */) override {
       C10_THROW_ERROR(Error, "FutureNCCL::markCompleted is not supported.");
     }
 
-    // Just returns NCCL collective's outputs after WorkNCCL's wait returns.
+    // Just returns FutureNCCL's value.
     at::IValue value() override {
-      work_->wait();
-      return outputs_;
+      return value_;
     }
 
     const at::IValue& constValue() override {
-      work_->wait();
-      return outputs_;
+      return value_;
```
We should probably just call `wait()` in here to be safe if the user doesn't call `wait()` themselves. This should be fine in our callbacks, since `.value()` is called inside our newly created stream, so `wait()` will block on the newly created stream and not the default stream.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -143,66 +144,129 @@ class ProcessGroupNCCL : public ProcessGroup {
+      // Set a default fut, this will be returned in case of an error in the
+      // callback. If the callback is executed without any errors, fut will be
+      // replaced by a FutureNCCL.
+      auto fut = c10::make_intrusive<Future>(type);
```
This is a little messy since you're creating a Future of default type and then replacing it with FutureNCCL inside the callback. You probably want to create `FutureNCCL` here and then call `markCompleted()` inside the lambda below. For this, you need another `FutureNCCL` constructor that doesn't take `value_` as input. Then, `markCompleted()` allows the user to set the value only if it is not set already. You can use `value_->isNone()` to check if it has been set or not.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -143,66 +144,129 @@ class ProcessGroupNCCL : public ProcessGroup {
+                  fut = c10::make_intrusive<FutureNCCL>(
+                      cb(), deviceIndex_, thenFutCudaEvent);
+                } catch (const std::exception& e) {
+                  fut->setError(e.what());
```
You probably need to override `setError` in FutureNCCL to ensure this is handled appropriately. Also, `wait()`, `value()` and `constValue()` need to look for an error being set.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -143,66 +144,129 @@ class ProcessGroupNCCL : public ProcessGroup {
+    // record the callback's stream.
+    void addCallbackWithStream(
+        std::function<void(void)> callback,
+        const c10::cuda::CUDAStream stream,
```
nit: `const c10::cuda::CUDAStream& stream`

sinannasir committed:
Update on "[NCCL] Changed FutureNCCL's then callback logic for better efficiency."


We realized that when we invoke a simple callback that divides the tensors by `world_size` after `allreduce`, the performance was almost 50% lower in terms of QPS compared to the case where a simple `allreduce` hook is used with no `then` callback.

@pritamdamania87 realized that we should run the callback in `then` on a different stream and synchronize the NCCL stream to that stream instead and not the default device stream. Now a new `FutureNCCL` is returned from `then` hold the return value of the callback and new cudaEvents that recorded the stream that guards this callback.. This new FutureNCCL actually synchronizes on the new stream we used for the callback when its wait() is called.

**Ongoing discussions:**
* Calling `fut.wait()` (instead of `fut.value()`) inside the callback of the allreduce then divide hook is okay.
* Calling `fut.wait()` if a new futureNCCL object is defined inside then callback is essential to get correct results. It could be fine in terms of performance since within the scope of then callback the default stream becomes the stream that guards the callback but not PyTorch's default device stream.

Differential Revision: [D23055807](https://our.internmc.facebook.com/intern/diff/D23055807/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] Changed FutureNCCL's then callback logic for better efficiency."


We realized that when we invoke a simple callback that divides the tensors by `world_size` after `allreduce`, the performance was almost 50% lower in terms of QPS compared to the case where a simple `allreduce` hook is used with no `then` callback.

@pritamdamania87 realized that we should run the callback in `then` on a different stream and synchronize the NCCL stream to that stream instead and not the default device stream. Now a new `FutureNCCL` is returned from `then` hold the return value of the callback and new cudaEvents that recorded the stream that guards this callback.. This new FutureNCCL actually synchronizes on the new stream we used for the callback when its wait() is called.

**Ongoing discussions:**
* Calling `fut.wait()` (instead of `fut.value()`) inside the callback of the allreduce then divide hook is okay.
* Calling `fut.wait()` if a new futureNCCL object is defined inside then callback is essential to get correct results. It could be fine in terms of performance since within the scope of then callback the default stream becomes the stream that guards the callback but not PyTorch's default device stream.

Differential Revision: [D23055807](https://our.internmc.facebook.com/intern/diff/D23055807/)

[ghstack-poisoned]

sinannasir commented:
> Can we describe the perf issue with the previous solution and why this commit fixes it (basically no longer blocking default stream on allreduce in then(), which eliminated overlap)? Would be useful context for those who come across this in the future.

I agree. I included additional small description in the documentation. I'll update the summary accordingly and give more details.

sinannasir committed:
Update on "[NCCL] Changed FutureNCCL's then callback logic for better efficiency."


We realized that when we invoke a simple callback that divides the tensors by `world_size` after `allreduce`, the performance was almost 50% lower in terms of QPS compared to the case where a simple `allreduce` hook is used with no `then` callback.

@pritamdamania87 realized that we should run the callback in `then` on a different stream and synchronize the NCCL stream to that stream instead and not the default device stream. Now a new `FutureNCCL` is returned from `then` hold the return value of the callback and new cudaEvents that recorded the stream that guards this callback.. This new FutureNCCL actually synchronizes on the new stream we used for the callback when its wait() is called.

**Ongoing discussions:**
* Calling `fut.wait()` (instead of `fut.value()`) inside the callback of the allreduce then divide hook is okay.
* Calling `fut.wait()` if a new futureNCCL object is defined inside then callback is essential to get correct results. It could be fine in terms of performance since within the scope of then callback the default stream becomes the stream that guards the callback but not PyTorch's default device stream.

Differential Revision: [D23055807](https://our.internmc.facebook.com/intern/diff/D23055807/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -154,57 +154,93 @@ class ProcessGroupNCCL : public ProcessGroup {
     }
 
     const at::IValue& constValue() override {
+      wait();
```
Add `TORCH_CHECK` here asserting that we have a value (hasValue)

sinannasir replied:
@pritamdamania87 If FutureNCCL doesn't have a value, there will be an error from the then callback and wait() will throw an error. Do we still need an additional assert here?

pritamdamania87 replied:
What if there is no `.then` callback? In that case, `wait()` wouldn't throw an error right?

sinannasir committed:
Update on "[NCCL] Changed FutureNCCL's then callback logic for better efficiency."


We realized that when we invoke a simple callback that divides the tensors by `world_size` after `allreduce`, the performance was almost 50% lower in terms of QPS compared to the case where a simple `allreduce` hook is used with no `then` callback.

@pritamdamania87 realized that we should run the callback in `then` on a different stream and synchronize the NCCL stream to that stream instead and not the default device stream. Now a new `FutureNCCL` is returned from `then` hold the return value of the callback and new cudaEvents that recorded the stream that guards this callback.. This new FutureNCCL actually synchronizes on the new stream we used for the callback when its wait() is called.

**Ongoing discussions:**
* Calling `fut.wait()` (instead of `fut.value()`) inside the callback of the allreduce then divide hook is okay.
* Calling `fut.wait()` if a new futureNCCL object is defined inside then callback is essential to get correct results. It could be fine in terms of performance since within the scope of then callback the default stream becomes the stream that guards the callback but not PyTorch's default device stream.

Differential Revision: [D23055807](https://our.internmc.facebook.com/intern/diff/D23055807/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] Changed FutureNCCL's then callback logic for better efficiency."


We realized that when we invoke a simple callback that divides the tensors by `world_size` after `allreduce`, the performance was almost 50% lower in terms of QPS compared to the case where a simple `allreduce` hook is used with no `then` callback.

The main problem was as we call `work.wait()` before invoking `then` callback, we were synchronizing `work`'s stream with the default PyTorch stream inside [`runHook`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L609) and stalling the backward computation.

In that PR, we ensure that FutureNCCL's `then` callback is not stalling the backward computation. Assuming single-process single-device, `FutureNCCL` gets a new stream from device's pool using `at::cuda::getStreamFromPool` to run `callback` and before invoking the `callback` inline it synchronizes `WorkNCCL`'s stream by callback's stream not the default stream.

Differential Revision: [D23055807](https://our.internmc.facebook.com/intern/diff/D23055807/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] Changed FutureNCCL's then callback logic for better efficiency."


We realized that when we invoke a simple callback that divides the tensors by `world_size` after `allreduce`, the performance was almost 50% lower in terms of QPS compared to the case where a simple `allreduce` hook is used with no `then` callback.

The main problem was as we call `work.wait()` before invoking `then` callback, we were synchronizing `work`'s stream with the default PyTorch stream inside [`runHook`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L609) and stalling the backward computation.

In that PR, we ensure that FutureNCCL's `then` callback is not stalling the backward computation. Assuming single-process single-device, `FutureNCCL` gets a new stream from device's pool using `at::cuda::getStreamFromPool` to run `callback` and before invoking the `callback` inline it synchronizes `WorkNCCL`'s stream by callback's stream not the default stream.

Differential Revision: [D23055807](https://our.internmc.facebook.com/intern/diff/D23055807/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] Changed FutureNCCL's then callback logic for better efficiency."


We realized that when we invoke a simple callback that divides the tensors by `world_size` after `allreduce`, the performance was almost 50% lower in terms of QPS compared to the case where a simple `allreduce` hook is used with no `then` callback.

The main problem was as we call `work.wait()` before invoking `then` callback, we were synchronizing `work`'s stream with the default PyTorch stream inside [`runHook`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L609) and stalling the backward computation.

In that PR, we ensure that FutureNCCL's `then` callback is not stalling the backward computation. Assuming single-process single-device, `FutureNCCL` gets a new stream from device's pool using `at::cuda::getStreamFromPool` to run `callback` and before invoking the `callback` inline it synchronizes `WorkNCCL`'s stream by callback's stream not the default stream.

Differential Revision: [D23055807](https://our.internmc.facebook.com/intern/diff/D23055807/)

[ghstack-poisoned]
